src/tls/Makefile.common

```diff
@@ -14,48 +37,82 @@ INCLUDE_PATHS= \
 endif
 
-CACHE_DIR=cache.$(FLAVOR)$(CONCRETE_FLAVOR)
-
 FSTAR=$(FSTAR_HOME)/bin/fstar.exe --cache_dir $(CACHE_DIR) --cache_checked_modules $(OTHERFLAGS)
 
-EXTRACT_DIR=extract/$(FLAVOR)$(CONCRETE_FLAVOR)
+VFLAGS?=--use_hints --use_hint_hashes --record_hints --warn_error -271
+
+HINTS_DIR  ?= .
+CACHE_DIR   = cache/$(FLAVOR)$(CONCRETE_FLAVOR)
+EXTRACT_DIR = extract/$(FLAVOR)$(CONCRETE_FLAVOR)
+DEP_FILE    = $(CACHE_DIR)/.depend
 
-ROOTS ?= TLS.fst QUIC.fst Test.Main.fst
 
-.depend.$(FLAVOR)$(CONCRETE_FLAVOR):
-	mkdir -p $(EXTRACT_DIR)
-	$(FSTAR) --odir $(EXTRACT_DIR) \
-		 --lax \
-		 --dep full \
-		 --extract $(EXTRACT) \
-		 $(addprefix --include , $(INCLUDE_PATHS)) \
-		 $(ROOTS) > $@
+# Project Files
 
-depend: .depend.$(FLAVOR)$(CONCRETE_FLAVOR)
+# Add more roots here!
+ROOTS ?= TLS.fst QUIC.fst Test.Main.fst 
 
--include .depend.$(FLAVOR)$(CONCRETE_FLAVOR)
+EVERYTHING=$(wildcard *.fst *.fsti Make* $(MITLS_HOME)/src/pki/* $(MITLS_HOME)/libs/ffi/* ideal-flags/* concrete-flags/* concrete-flags/$(FLAVOR)$(CONCRETE_FLAVOR)/*)
 
-%.checked.lax:
-	$(FSTAR) --lax $(addprefix --include , $(INCLUDE_PATHS)) $<
-	touch $@
+
+# Production Rules
+
+$(DEP_FILE): $(EVERYTHING) 
+ifdef VERBOSE
+	@echo "\033[1;33mComputing $(FLAVOR)$(CONCRETE_FLAVOR) dependencies ...\033[;37m"
+endif
+	@mkdir -p $(EXTRACT_DIR)
+	@mkdir -p $(CACHE_DIR)
+	@mkdir -p $(HINTS_DIR)
+	$(subst \\,, \
+		$(FSTAR) --cache_dir $(CACHE_DIR) --odir $(EXTRACT_DIR) --extract $(EXTRACT) \
+			--dep full \
+			$(addprefix --include , $(INCLUDE_PATHS)) \
+			$(ROOTS) > $(DEP_FILE) \
+	)
+
+depend: $(DEP_FILE)
+
+-include $(DEP_FILE)
+
+# the touch is needed because make may decide to invoke fstar on this
+# target because the timestamp on, say, the .fst file changed, and
+# then fstar may detect that there was no change in content and not
+# re-emit the .fst.checked file. Adding the touch ensures that the
+# timestamp on the .checked file is updated to avoid make running this
+# rule again.
+
+%.checked: | $(DEP_FILE)
+ifdef VERBOSE
+	@echo "\033[1;35m=== Producing $@ ...\033[;37m"
+endif
+	$(subst \\,, \
+		$(FSTAR) $(VFLAGS) --hint_file $(HINTS_DIR)/$<.hints $(addprefix --include , $(INCLUDE_PATHS)) $< \
+	)
+	@touch $@
 
 $(EXTRACT_DIR)/%.$(EXTENSION):
-	$(FSTAR) \
-		--codegen $(FLAVOR) \
-		--odir $(EXTRACT_DIR) \
-		--lax \
-		$(addprefix --include , $(INCLUDE_PATHS)) \
-		--extract_module $(basename $(notdir $(subst .checked.lax,,$<))) \
-		$(subst $(CACHE_DIR)/,,$(subst .checked.lax,, $<))
+ifdef VERBOSE
+	@echo "\033[1;32m=== Extracting $@ ...\033[;37m"
+endif
+	$(subst \\,, \
+		$(FSTAR) $(VFLAGS) \
+			--codegen $(FLAVOR) \
+			--cache_dir $(CACHE_DIR) \
+			--hint_file $(HINTS_DIR)/$(subst .lax,,$(subst .checked,,$<)).hints \
+			--odir $(EXTRACT_DIR) \
+			$(addprefix --include , $(INCLUDE_PATHS)) \
+			--extract_module $(basename $(notdir $(subst .lax,,$(subst .checked,,$<)))) \
+			$(subst $(CACHE_DIR)/,,$(subst .lax,,$(subst .checked,,$<))) \
```

Comment:
are all these substitutions necessary? is it to deal with both .lax and .checked.lax files?

Reply:
Yeah, to deal with each of them separately. Now that we're not using .lax files anymore, we could remove that half.